File: efficient-server-cpp/src/kdtree.cpp
#include <cstdint>
#include <iostream>
#include "stdio.h"
#include "kdtree.h"
#include <queue>
#include <algorithm>
#include <execution>
#include <numeric>
#include <cstdint>
#include <limits>
#include <limits>

using namespace std;
using weight_id = pair<uint64_t,int64_t>; /// {distance, id}

KDTree tree;

int64_t Vertex::id_count = 0;
const double FIFTY_CM = 500.0; /// in mm
const double FIFTY_CM_SQRD = 250000.0; /// in mm^2
const int DIMENSIONS = 2; /// 2D tree
const int NOT_FOUND = -1;


KDTree::KDTree() {
}


bool KDTree::inProximity(Point a, int64_t nodeId) {
    int64_t nodeX = vertices[nodeId].coords[0];
    int64_t nodeY = vertices[nodeId].coords[1];
    auto dx = (double) (a.coords[0] - nodeX);
    auto dy = (double) (a.coords[1] - nodeY);
    return dx * dx + dy * dy <= FIFTY_CM_SQRD;
}


uint64_t KDTree::oneToAll(int64_t src) {
    priority_queue<weight_id, vector<weight_id>, greater<weight_id>> PQ; /// Min heap
    vector<uint64_t> dist_from_src(vertices.size(), UINT64_MAX);
    dist_from_src[src] = 0;

    PQ.emplace(0, src);

    // Dijkstra's algorithm to find shortest paths
    while (!PQ.empty()) {
        // Extract the node with the minimum distance from the priority queue
        weight_id curr = PQ.top(); PQ.pop();
        auto curr_dist = curr.first;
        auto curr_id = curr.second;

        // Iterate over the nbrs of the current node
        for (auto nbr : vertices[curr_id].nbrs) {
            auto nbr_id = nbr.first;
            weight_count pair = nbr.second;
            uint64_t new_dist = curr_dist + pair.first / pair.second; /// average distance
            // If the new path is not shorter, skip it
            if (new_dist >= dist_from_src[nbr_id])
                continue;

            PQ.emplace(new_dist, nbr_id);
            dist_from_src[nbr_id] = new_dist;
        }
    }

    // parallel sum
    uint64_t sum = std::transform_reduce(std::execution::par, dist_from_src.begin(), dist_from_src.end(), 0ULL,
                                         std::plus<>(), [](uint64_t dist) {
                return dist == UINT64_MAX ? 0ULL : dist;
            });

    return sum;
}



uint64_t KDTree::oneToOne(int64_t src, int64_t dst) {

    vector<uint64_t> dist_from_src(vertices.size(), UINT64_MAX);
    dist_from_src[src] = 0;

    vector<int64_t> prev(vertices.size(), NOT_FOUND);
    vector<bool> closed(vertices.size(), false);

    priority_queue<weight_id, vector<weight_id>, greater<weight_id>> PQ;
    PQ.emplace(0, src);


    while (!PQ.empty()) {

        if (PQ.top().second != dst) {
            weight_id curr = PQ.top();
            PQ.pop();

            for_each(execution::par_unseq, vertices[curr.second].nbrs.begin(), vertices[curr.second].nbrs.end(), [&](auto& nbr) {
                int64_t nbr_id = nbr.first;
                weight_count pair = nbr.second;
                int new_dist = curr.first + pair.first / pair.second;
                if (new_dist < dist_from_src[nbr_id]) {
                    PQ.emplace(new_dist, nbr_id);
                    dist_from_src[nbr_id] = new_dist;
                    prev[nbr_id] = curr.second;
                }
            });
        } else break;
    }
    return dist_from_src[dst];
}


int64_t KDTree::insertRec(int64_t nodeId, Point point, int depth) {
    if (!vertices.empty()) {
        int cd = depth % DIMENSIONS;
        /// current node is lesser than the node we are trying to add
        bool curr_is_lesser = point.coords[cd] < vertices[nodeId].coords[cd];
        int64_t nextNodeId = curr_is_lesser ? vertices[nodeId].left : vertices[nodeId].right;

        if (nextNodeId != NOT_FOUND) {
            return insertRec(nextNodeId, point, ++depth);
        } else {
            vertices.emplace_back(point);
            curr_is_lesser ? vertices[nodeId].left = vertices.back().id : vertices[nodeId].right = vertices.back().id;
            return vertices.back().id;
        }
    } else {
        vertices.emplace_back(point);
        return vertices.back().id;

    }
}

int64_t KDTree::insert(Point point) {
    return insertRec(0, point, 0);
}


int64_t KDTree::searchTree(int64_t nodeId, Point point, int depth) {
    if (nodeId != NOT_FOUND && !vertices.empty()) {

        if (!inProximity(point, nodeId)) {

            int cd = depth % DIMENSIONS;
            int64_t ret = NOT_FOUND;
            if (point.coords[cd] - FIFTY_CM <= vertices[nodeId].coords[cd])
                ret = searchTree(vertices[nodeId].left, point, depth + 1);

            if (ret == NOT_FOUND && point.coords[cd] + FIFTY_CM >= vertices[nodeId].coords[cd])
                ret = searchTree(vertices[nodeId].right, point, depth + 1);

            return ret;
        } else return nodeId;
    } else return NOT_FOUND;
}

int64_t KDTree::searchProximity(Point point) {
    return searchTree(0, point, 0);
}

File: esw-eff-server-java/src/main/java/server/KDTree.java
package server;

import cz.cvut.fel.esw.server.proto.Location;

import java.util.*;

public class KDTree {
    private static final double FIFTY_CM = 500.0; // in mm
    private static final double FIFTY_CM_SQRD = 250000.0; // in mm^2
    private static final int DIMENSIONS = 2; // 2D tree
    private static final int NOT_FOUND = -1;

    public List<Node> nodes;

    public KDTree() {
        nodes = new ArrayList<>();
    }

    public long insert(Point point) {
        if (nodes.isEmpty()) {
            nodes.add(new Node(point));
            return nodes.get(0).id;
        }

        int depth = 0;
        Node currentNode = nodes.get(0);
        while (true) {
            int cd = depth % DIMENSIONS;
            boolean currIsLesser = point.coords[cd] < currentNode.coords[cd];
            long nextNodeId = currIsLesser ? currentNode.left : currentNode.right;

            if (nextNodeId == NOT_FOUND) {
                nodes.add(new Node(point));
                Node newNode = nodes.get(nodes.size() - 1);
                if (currIsLesser)
                    currentNode.left = newNode.id;
                else
                    currentNode.right = newNode.id;
                return newNode.id;
            } else {
                currentNode = nodes.get((int) nextNodeId);
                depth++;
            }
        }
    }



    public long findNear(Point point) {
        return searchTree(0, point, 0);
    }

    private long searchTree(long nodeId, Point point, int depth) {
        while (nodeId != NOT_FOUND && !nodes.isEmpty()) {
            if (inProximity(point, nodeId))
                return nodeId;

            int cd = depth % DIMENSIONS;
            Node currentNode = nodes.get((int) nodeId);

            long ret = NOT_FOUND;
            if (point.coords[cd] - FIFTY_CM <= currentNode.coords[cd])
                ret = searchTree(currentNode.left, point, depth + 1);

            if (ret == NOT_FOUND && point.coords[cd] + FIFTY_CM >= currentNode.coords[cd])
                ret = searchTree(currentNode.right, point, depth + 1);

            return ret;
        }

        return NOT_FOUND;
    }



    public long oneToAll(long src) {
        PriorityQueue<DistancePair> priorityQueue = new PriorityQueue<>(Comparator.comparingLong(dp -> dp.distance));
        long[] distFromSrc = new long[nodes.size()];
        Arrays.fill(distFromSrc, Long.MAX_VALUE);
        distFromSrc[(int) src] = 0;

        priorityQueue.add(new DistancePair(0, src));

        while (!priorityQueue.isEmpty()) {
            DistancePair current = priorityQueue.poll();
            long currentDist = current.distance;
            long currentId = current.nodeId;

            Node currentNode = nodes.get((int) currentId);
            for (Map.Entry<Long, Edge> entry : currentNode.nbrs.entrySet()) {
                long nbrId = entry.getKey();
                Edge nbr = entry.getValue();
                long newDist = currentDist + nbr.dist / nbr.count;
                if (newDist < distFromSrc[(int) nbrId]) {
                    priorityQueue.add(new DistancePair(newDist, nbrId));
                    distFromSrc[(int) nbrId] = newDist;
                }
            }
        }

        long sum = 0;
        for (long dist : distFromSrc) {
            if (dist != Long.MAX_VALUE)
                sum += dist;
        }

        return sum;
    }

    public long oneToOne(long src, long dst) {
        long[] distFromSrc = new long[nodes.size()];
        Arrays.fill(distFromSrc, Long.MAX_VALUE);
        distFromSrc[(int) src] = 0;

        long[] prev = new long[nodes.size()];
        Arrays.fill(prev, NOT_FOUND);

        PriorityQueue<DistancePair> priorityQueue = new PriorityQueue<>(Comparator.comparingLong(dp -> dp.distance));
        priorityQueue.add(new DistancePair(0, src));

        while (!priorityQueue.isEmpty()) {
            DistancePair current = priorityQueue.poll();
            if (current.nodeId == dst)
                break;

            Node currentNode = nodes.get((int) current.nodeId);
            for (Map.Entry<Long, Edge> entry : currentNode.nbrs.entrySet()) {
                long nbrId = entry.getKey();
                Edge nbr = entry.getValue();
                long newDist = current.distance + nbr.dist / nbr.count;
                if (newDist < distFromSrc[(int) nbrId]) {
                    priorityQueue.add(new DistancePair(newDist, nbrId));
                    distFromSrc[(int) nbrId] = newDist;
                    prev[(int) nbrId] = current.nodeId;
                }
            }
        }

        return distFromSrc[(int) dst];
    }

    public boolean inProximity(Point a, long nodeId) {
        long nodeX = nodes.get((int) nodeId).coords[0];
        long nodeY = nodes.get((int) nodeId).coords[1];
        double dx = a.coords[0] - nodeX;
        double dy = a.coords[1] - nodeY;
        return dx * dx + dy * dy <= FIFTY_CM_SQRD;
    }

    public static class Node {
        long id;
        static long idCount;
        long[] coords;
        long left;
        long right;
        Map<Long, Edge> nbrs;

        Node(Point point) {
            coords = new long[2];
            coords[0] = point.coords[0];
            coords[1] = point.coords[1];
            left = -1;
            right = -1;
            id = idCount++;
            nbrs = new HashMap<>();
        }

        void addNeighbor(int dist, long neighborId) {
            if (nbrs.containsKey(neighborId)) {
                Edge edge = nbrs.get(neighborId);
                edge.dist += dist;
                edge.count++;
            } else {
                nbrs.put(neighborId, new Edge(dist, 1));
            }
        }
    }

    private static class Edge {
        long dist;
        int count;

        Edge(long dist, int count) {
            this.dist = dist;
            this.count = count;
        }
    }

    public static class Point {
        long[] coords;

        Point(Location location) {
            coords = new long[2];
            coords[0] = location.getX();
            coords[1] = location.getY();
        }
    }

    private static class DistancePair {
        long distance;
        long nodeId;

        DistancePair(long distance, long nodeId) {
            this.distance = distance;
            this.nodeId = nodeId;
        }
    }
}
